src/mpi4py/MPI.src/asdlpack.pxi

```diff
@@ -208,40 +231,61 @@ cdef int Py_GetDLPackBuffer(object obj, Py_buffer *view, int flags) except -1:
         raise NotImplementedError("dlpack: missing support")
 
     device_type, device_id = dlpack_device()
-    if device_type == kDLCPU:
-        capsule = dlpack()
-    else:
-        capsule = dlpack(stream=-1)
-        <void> device_id  # unused
-    if not PyCapsule_IsValid(capsule, b"dltensor"):
-        raise BufferError("dlpack: invalid capsule object")
+    <void> device_id  # unused
 
-    managed = <DLManagedTensor*> PyCapsule_GetPointer(capsule, b"dltensor")
-    dltensor = &managed.dl_tensor
+    try:  # DLPack v1.0+
+        if device_type == kDLCPU:
+            capsule = dlpack(max_version=(version, 0), copy=False)
+        else:
+            capsule = dlpack(stream=-1, max_version=(version, 0), copy=False)
+    except TypeError:  # DLPack v0.x
+        version = 0
+        capsulename = b"dltensor"
+        usedcapsulename = b"used_dltensor"
+        if device_type == kDLCPU:
+            capsule = dlpack()
+        else:
+            capsule = dlpack(stream=-1)
+
+    if not PyCapsule_IsValid(capsule, capsulename):
+        raise BufferError("dlpack: invalid capsule object")
+    pointer = PyCapsule_GetPointer(capsule, capsulename)
+    if version >= 1:
+        managed1 = <DLManagedTensorVersioned*> pointer
```

Comment:
@leofang Should we add a sanity check to make sure that the (major) version value is the expected one (>=1), or should we just trust the producer?

Reply:
What kind of sanity check do you have in mind? I don't see how we can avoid a type cast here?

Reply:
I meant just something like the following assert-based check:

```python
if version >= 1:
        managed1 = <DLManagedTensorVersioned*> pointer
        assert managed1.version.major >= 1
```

Reply:
Doesn't hurt to add if you like.